### ISSUE TITLE
DBeaver incorrectly parses Snowflake stored procedure that contains the text: //end

### ISSUE THREAD

**Issue by mleehayes** [dbeaver/dbeaver#13490]
#### System information: 
- Operating system (distribution) and version: Windows 10 Version 1909 (OS Build 18363.1679)
- DBeaver version: 21.1.4.202108020335
- Additional extensions: ??

#### Connection specification:
- Database name and version: Snowflake
- Driver name:  JDBC 3.12.15
- Do you use tunnels or proxies (SSH, SOCKS, etc)? 

#### Describe the problem you're observing:
When I attempt to create a stored procedure that contains the text "//end" (without double-quotes),  I get a message indicating that DBeaver has incorrectly parsed the SQL statement.  In the example below,  if I have the cursor in the stored procedure section and press CTL + ENTER, I get the message:  "SQL Error [8] [0A000]: Actual statement count 2 did not match the desired statement count 1." even though there's only one statement to execute at the current cursor position.

If I don't have a SQL statement after the stored procedure statement,  I don't get the error.

Note that the actual symptoms are different if the stored procedure is more complex.  In the below example,  if the body of the stored procedure is changed to:  
  x = 1 + 2
  //end 
  return "fred";

Then I get a different error:
  SQL Error [1003] [42000]: SQL compilation error:
  parse error line 14 at position 13 near '<EOF>'.
  syntax error line 8 at position 1 unexpected '$'.

but I think the underlying issue is the same.

#### Steps to reproduce, if exist:
Execute the following statement in DBeaver and execute:

===============================================
-- CONNECTION: name=uat_trst_ent_db_dedupe_role
CREATE OR REPLACE PROCEDURE uat_trst_ent_db_preprd.utility.r626992_test_sproc()

RETURNS VARCHAR(16777216)
LANGUAGE JAVASCRIPT

EXECUTE AS CALLER
AS
$$

return "fred";
//end 
$$
;
select current_client()
;
===============================================

#### Include any warning/errors/backtraces from the logs

Extract from log:
!ENTRY org.jkiss.dbeaver.model 4 0 2021-08-10 15:40:44.632
!MESSAGE SQL Error [8] [0A000]: Actual statement count 2 did not match the desired statement count 1.
!SUBENTRY 1 org.jkiss.dbeaver.model 4 0 2021-08-10 15:40:44.632
!MESSAGE SQL Error [8] [0A000]: Actual statement count 2 did not match the desired statement count 1.
!SUBENTRY 2 org.jkiss.dbeaver.model 4 0 2021-08-10 15:40:44.632
!MESSAGE Actual statement count 2 did not match the desired statement count 1.
!STACK 0
net.snowflake.client.jdbc.SnowflakeSQLException: Actual statement count 2 did not match the desired statement count 1.
	at net.snowflake.client.jdbc.SnowflakeUtil.checkErrorAndThrowExceptionSub(SnowflakeUtil.java:124)
	at net.snowflake.client.jdbc.SnowflakeUtil.checkErrorAndThrowException(SnowflakeUtil.java:64)
	at net.snowflake.client.core.StmtUtil.pollForOutput(StmtUtil.java:434)
	at net.snowflake.client.core.StmtUtil.execute(StmtUtil.java:338)
	at net.snowflake.client.core.SFStatement.executeHelper(SFStatement.java:501)
	at net.snowflake.client.core.SFStatement.executeQueryInternal(SFStatement.java:229)
	at net.snowflake.client.core.SFStatement.executeQuery(SFStatement.java:167)
	at net.snowflake.client.core.SFStatement.execute(SFStatement.java:749)
	at net.snowflake.client.jdbc.SnowflakeStatementV1.executeInternal(SnowflakeStatementV1.java:292)
	at net.snowflake.client.jdbc.SnowflakeStatementV1.execute(SnowflakeStatementV1.java:342)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:330)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:130)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:513)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$0(SQLQueryJob.java:444)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:171)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:431)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:816)
	at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:3435)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:118)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:171)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:116)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:4711)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

**Comment by Matvey16** [comment]
@mleehayes 
Hello, I tried reproducing your issue and seems like expected snowflake behavior. How Dbeaver parses your script is influenced by it's structure and SQL processing properties. Can you show me you r properties for SQL processing

![dbeaver_bYj4YJ3I6s](https://user-images.githubusercontent.com/82543000/129889208-220bf160-94b1-4abe-a9b2-ba62270c3d90.png)

**Comment by mleehayes** [comment]
Here's the screenshot:

![SQL_processing](https://user-images.githubusercontent.com/73184896/130254770-aca46d75-963a-49ff-8000-8c50fc2d5e45.jpg)

I guess I wasn't clear on the issue.  Snowflake supports JavaScript as it's stored procedure language.  The forward slash characters are line comments in the stored procedure and are ignored by Snowflake.  The problem as I see it is that DBeaver is interpreting the //end statement as the end of the stored procedure,  It's not the end.

**Comment by Matvey16** [comment]
@mleehayes

> When I attempt to create a stored procedure that contains the text "//end" (without double-quotes), I get a message indicating that DBeaver has incorrectly parsed the SQL statement. In the example below, if I have the cursor in the stored procedure section and press CTL + ENTER, I get the message: "SQL Error [8] [0A000]: Actual statement count 2 did not match the desired statement count 1." even though there's only one statement to execute at the current cursor position.
> 
> If I don't have a SQL statement after the stored procedure statement, I don't get the error.

This is expected Snowflake behavior, because you can't run two statements in one query, you need to either use Alt+X to run the whole script or first execute `alter session set MULTI_STATEMENT_COUNT=0;` to disable the lock on multi statement, but be aware of the vulnerability to SQL injections.

And I was not able to reproduce the issue, can you look and verify that I did everything like you

![nhmY7N6C82](https://user-images.githubusercontent.com/82543000/130735431-4f391293-317b-4fac-92bd-39a582361308.gif)

 
![dbeaver_21WD3WbWiV](https://user-images.githubusercontent.com/82543000/130735488-a033b34a-d18c-4037-8314-b8340a252271.png)

**Comment by charlesHenryPATARD** [comment]
Hi,
I'm facing the same issue.

My configuration : 

SGBD : Ingres
DBeaver version :  21.5.5
OS : Linux CentOS

I changed my SQL processing settings as mentioned in previous post by @Matvey16 like this : 

Uncheck : Ignore native delimiter
Uncheck : Blank line...
Check : Remove trailing query delimiter

My Procedure is : 

`CREATE PROCEDURE p_test_chp
	(id INTEGER NOT NULL) as
	
declare
	id_level INT;
BEGIN
	SELECT id_level into :id_level FROM philippe.accelerator_levels where id_level = :id;
end;`

I finally found a solution : Remove the ; after end key word.

**Comment by mleehayes** [comment]
I agree you weren't able to reproduce the behavior.  Weird.  My settings appear to be the same as yours.  I guess we close the issue as unreproduceable.  

Thanks for looking into the issue.

**Comment by mleehayes** [comment]
Matvey16,  I just ran across a potential solution while working on a Snowflake UDF.  In my connection settings, in the Snowflake tab,  I had the following selected:  
![image](https://user-images.githubusercontent.com/73184896/131581914-bb12f2b5-7807-4ccf-86a8-7b5e06ca0c5a.png)

If I changed the setting from "Code block" to "String literal",  it solved the issue I was having with the UDF.  I haven't checked the stored procedure, but my guess is that it will fix it also.
